Source/WebKit/UIProcess/EnhancedSecurityTracking.cpp

```diff
@@ -0,0 +1,243 @@
+        }
+    }
+}
+
+void EnhancedSecurityTracking::trackSameSiteNavigation(API::Navigation& navigation)
+{
+    bool isHttps = navigation.currentRequest().url().protocolIs("https"_s);
+
+    if (enhancedSecurityReason() == EnhancedSecurityReason::InsecureProvisional && isHttps) {
+        reset();
+
+        auto domain = RegistrableDomain { navigation.currentRequest().url() };
+        if (enabledSitesMap().contains(domain)) {
+            if (enabledSitesMap().get(domain) == EnhancedSecurityReason::InsecureProvisional)
+                enabledSitesMap().remove(domain);
+        }
+    }
+}
+
+bool EnhancedSecurityTracking::enableIfRequired(API::Navigation& navigation)
+{
+    auto currentRequestURL = navigation.currentRequest().url();
+
+    if (currentRequestURL.protocolIs("http"_s) && !WebCore::isLocalIPAddressSpace(currentRequestURL)) {
+        enableFor(EnhancedSecurityReason::InsecureProvisional, navigation);
+        return true;
+    }
+
+    return false;
+}
+
+void EnhancedSecurityTracking::handleBackForwardNavigation(API::Navigation& navigation)
+{
+    EnhancedSecurity priorState = navigation.targetItem() ? navigation.targetItem()->enhancedSecurity() : EnhancedSecurity::Disabled;
+
+    if (priorState == EnhancedSecurity::Disabled) {
+        if (m_activeState != ActivationState::None)
+            makeDormant();
+    } else
+        enableFor(reasonForEnhancedSecurity(priorState), navigation);
+}
+
+void EnhancedSecurityTracking::trackNavigation(API::Navigation& navigation, CompletionHandler<void()>&& reply)
+{
+    auto lastNavigationAction = navigation.lastNavigationAction();
+
+    bool isBackForward = lastNavigationAction && lastNavigationAction->navigationType == NavigationType::BackForward;
+    bool isReload = lastNavigationAction && lastNavigationAction->navigationType == NavigationType::Reload;
+    bool isInitialUIDriven = navigation.isRequestFromClientOrUserInput() && !navigation.currentRequestIsRedirect();
+
+    if (isBackForward) {
+        handleBackForwardNavigation(navigation);
+        return reply();
+    }
+
+    if (m_activeState != ActivationState::None && isInitialUIDriven && !isReload)
+        reset();
+
+    if (m_activeState != ActivationState::Active && enableIfRequired(navigation))
+        return reply();
+
+    if (m_activeState == ActivationState::Active
+        && m_activeReason == EnhancedSecurityReason::InsecureProvisional) {
+        if (!m_initialProtectedDomain.matches(navigation.currentRequest().url()))
+            trackChangingSiteNavigation(navigation);
+        else
+            trackSameSiteNavigation(navigation);
+    }
+
+    if (m_activeState == ActivationState::None)
+        return reply();
+
+    if (sitePreviouslyUsedEnhancedSecurity(navigation)) {
+        if (m_activeState == ActivationState::Dormant)
+            makeActive();
+
+        ASSERT(m_activeState == ActivationState::Active);
+        return reply();
+    }
+
+    hasMeaningfulSiteInteraction(navigation, [
```

Comment:
Instead of calling this after we receive the policy response, we should prefer calling this in parallel. You can see an example of this in the safe browsing check beginning in `WebPageProxy::decidePolicyForNavigationAction` (`beginSafeBrowsingCheck`). `WebFramePolicyListenerProxy` will block the callback response until we get all of the responses that we need.